server/src/domain/storage/storage.core.ts

```diff
@@ -140,29 +143,74 @@ export class StorageCore {
 
     let move = await this.moveRepository.getByEntity(entityId, pathType);
     if (move) {
+      console.log('Found existing move record for', entityId, pathType, oldPath, newPath);
       this.logger.log(`Attempting to finish incomplete move: ${move.oldPath} => ${move.newPath}`);
       const oldPathExists = await this.repository.checkFileExists(move.oldPath);
       const newPathExists = await this.repository.checkFileExists(move.newPath);
-      const actualPath = newPathExists ? move.newPath : oldPathExists ? move.oldPath : null;
+      const actualPath = oldPathExists ? move.oldPath : newPathExists ? move.newPath : null;
       if (!actualPath) {
         this.logger.warn('Unable to complete move. File does not exist at either location.');
         return;
       }
 
-      this.logger.log(`Found file at ${actualPath === move.oldPath ? 'old' : 'new'} location`);
+      const fileAtNewLocation = actualPath === move.newPath;
+      this.logger.log(`Found file at ${fileAtNewLocation ? 'new' : 'old'} location`);
+
+      if (fileAtNewLocation) {
+        if (!(await this.isNewPathSameSize(move.oldPath, move.newPath, assetInfo))) {
+          this.logger.fatal(`Skipping move due to file size mismatch, old file is missing and new file is different`);
+          return;
+        }
+      }
 
       move = await this.moveRepository.update({ id: move.id, oldPath: actualPath, newPath });
     } else {
       move = await this.moveRepository.create({ entityId, pathType, oldPath, newPath });
     }
 
+    if (pathType === AssetPathType.ORIGINAL && !assetInfo) {
+      this.logger.warn(`Unable to complete move. Missing asset info for ${entityId}`);
+      return;
+    }
+
     if (move.oldPath !== newPath) {
-      await this.repository.moveFile(move.oldPath, newPath);
+      try {
+        this.logger.debug(`Attempting to rename file: ${move.oldPath} => ${newPath}`);
+        await this.repository.rename(move.oldPath, newPath);
+      } catch (err: any) {
+        if (err.code !== 'EXDEV') {
+          this.logger.warn(
+            `Unable to complete move. Error renaming file with code ${err.code} and message: ${err.message}`,
+          );
+          return;
+        }
+        this.logger.debug(`Unable to rename file. Falling back to copy, verify and delete`);
+        await this.repository.copyFile(move.oldPath, newPath);
+
+        if (!(await this.isNewPathSameSize(move.oldPath, newPath, assetInfo))) {
+          this.logger.warn(`Skipping move due to file size mismatch`);
+          return;
+        }
+
+        await this.repository.unlink(move.oldPath);
```

Comment:
What if deleting the file fails?

Reply:
I think at that point we leave it to become orphaned? Would we really want to fail the entire move for that? I guess we should at least try catch around that though

Reply:
Yeah that's what I was thinking.